dcap/types/src/quote3.rs

```diff
@@ -44,34 +77,217 @@ impl<T: AsRef<[u8]>> Quote3<T> {
 
         data.ct_eq(report_data.as_ref()).into()
     }
+
+    /// Try to get a [`Quote3`] from `bytes`
+    ///
+    /// This will ensure `bytes` is for the correct quote type and that it's
+    /// large enough to represent the quote.
+    ///
+    /// # Arguments:
+    /// * `bytes` - The bytes to interpret as a [`Quote3`]
+    ///
+    /// # Errors:
+    /// * [`Error::InvalidInputLength`] if the length of `bytes` is not large
+    ///   enough to represent the [`Quote3`].
+    /// * [`Error::InvalidVersion`] if the `bytes` is for a different quote
+    ///   version.
+    fn try_from_bytes(bytes: T) -> Result<Self> {
+        let ref_bytes = bytes.as_ref();
+        let bytes_length = ref_bytes.len();
+        if bytes_length < MIN_QUOTE_SIZE {
+            return Err(Error::InvalidInputLength {
+                required_size: MIN_QUOTE_SIZE,
+                actual_size: bytes_length,
+            });
+        }
+
+        // This shouldn't fail since we checked for `MIN_QUOTE_SIZE` above.
+        let version = u16_from_bytes(ref_bytes, 0)? as usize;
+        if version != 3 {
+            return Err(Error::InvalidVersion(version));
+        }
+
+        let auth_data = AuthenticationData::try_from(&bytes.as_ref()[AUTH_DATA_OFFSET..]).map_err(
+            |e| match e {
```

Comment:
Not fond of the similarity between this closer and the one below.
I could make it a function and pass in the "base size" but felt it might be a bit of an indirection.

Reply:
Implemented a method on the Error type. I intentionally left off the inline:
- thinking we should leverage the compiler's heuristics.  
- thinking, IFF the compiler fails the optimal choice, optimizing the error case probably isn't a priority.